### ISSUE TITLE
Process which uses zfs hang, exporting pool caused REFERENCE_BY_POINTER BSOD

### ISSUE THREAD

**Issue by GeorgeGhost** [openzfsonwindows/ZFSin#261]
Ways to reproduce:

- Run BeamNG drive from steam library on zfs pool in Windows(steamlibrary chmoded (chmod 777) in linux)
- BeamNG hangs on loading terrain
- Closing it helps to remove game gui, but leaves unkillable process behind
- Running zpool export pool causes REFERENCE_BY_POINTER BSOD

Not sure if those 2 are separate issues or one originating from the other. But BeamNG runs fine from non-zfs partition.

**Comment by lundman** [comment]
It sounds like the hang is the issue, since that holds a reference open, you can not unmount (or export) and then it all falls apart when it tries to.

If I were to replicate it, I would get to the hang, then run windbg (devstudio attach) and dump all stacks to see where it is stuck. 
( `.logopen c:\src\stacks.txt ; !stacks 2;  .logclose `)

**Comment by GeorgeGhost** [comment]
The problem also occurs in TESV, both close the GUI when you force close them, but leave a process running, when attaching in windbg, its console closes immediately after opening. 
This issue seems to be related to ZFSin, as when launching those games from the different drive on windows or running from the same ZFS pool on Linux the issue does not occur and games behave normaly.

**Comment by GeorgeGhost** [comment]
Ruling out lz4 compressin being a problem as the issue occurs on both compressed and uncompressed datasets

**Comment by GeorgeGhost** [comment]
Tried different windows installation the issue still exists

**Comment by GeorgeGhost** [comment]
Done further testing and the issue seems to only occur on datasets, creating a ZVOL with NTFS partition showed no problems in games.

**Comment by lundman** [comment]
Yes, it is more likely to be related to how the memory-mapped code (doesn't) works, I suspect it isn't quite 100% there. But installing steam, and a game, is pretty difficult reproduce steps, I'm hoping to find a smaller (free) program that exhibits issues that we can debug with.

**Comment by GeorgeGhost** [comment]
The latest commits seemed to fix the issue partially, I can now get into the game and play it for a few minutes before the application hangs.
